Learn tidyverse with AI tools post (#719)
tidyverse/tidyverse.org: content/blog/learn-tidyverse-ai/index.Rmd

---
output: hugodown::hugo_document

slug: learn-tidyverse-ai
title: Learning the tidyverse with the help of AI tools
date: 2025-04-04
author: Mine Çetinkaya-Rundel
description: >
    Tips and recommendations for learning the tidyverse with AI tools.

photo:
  url: https://unsplash.com/photos/zwd435-ewb4
  author: Andrea De Santis

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [learn] 
tags: [ai]
---

```{=html}
<!--
TODO:
* [ ] Look over / edit the post's title in the yaml
* [ ] Edit (or delete) the description; note this appears in the Twitter card
* [ ] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [ ] Find photo & update yaml metadata
* [ ] Create `thumbnail-sq.jpg`; height and width should be equal
* [ ] Create `thumbnail-wd.jpg`; width should be >5x height
* [ ] `hugodown::use_tidy_thumbnails()`
* [ ] Add intro sentence, e.g. the standard tagline for the package
* [ ] `usethis::use_tidy_thanks()`
-->
```

As an educator who teaches data science with R, I have LOTS of opinions about using artificial intelligence (AI) tools when learning R.

> But I will keep this post the use of generative AI tools, like Chat GPT, in learning R, and specifically learning to do data science with R and the tidyverse.[^1]

Reviewer: ```suggestion
But I will keep this post to the use of generative AI tools, like ChatGPT, in learning R, and specifically learning to do data science with R and the tidyverse.[^1]
```

I'll first walk you through three case studies that demonstrate what asking for help from a generative AI tool looks like.
Then, I'll wrap up the post with some tips and good practices for getting the most out of help provided by these tools in the context of writing tidyverse code.

Before we get started, though, it's important to note that knowledge bases of Large Language Models (LLMs) that underlie popular generative AI tools are private and lack transparency.
There are important societal concerns about the fairness of equitable access to these tools.
It's unclear how developers or users of these models can be held accountable.
Additionally, there is an environmental impact to using these tools.

## Case study 1: Reshaping and plotting

The first case study features an example on reshaping and plotting the `billboard` data frame and it's directly out of [R for Data Science](https://r4ds.hadley.nz/).
In fact, it's [worked out in the book](https://r4ds.hadley.nz/data-tidy.html#sec-billboard).
Here is the initial prompt:

> Use the billboard dataset in the tidyr package to create a visualization of rank vs. week number for each song in the dataset.

> And the following is what [Chat GPT](https://chatgpt.com/) by Open AI suggests as the first answer.

Reviewer: ```suggestion
And the following is what [ChatGPT](https://chatgpt.com/) by OpenAI suggests as the first answer:
```

Reviewer: won't suggest the Chat GPT -> ChatGPT edit elsewhere in the post, but there are a few more :)

Reviewer: Recommending a colon just as that seems to be the punctuation used in similar situations in this doc.

Reviewer: Could we include a reference to the version of the model used in the chat the first time it's reference in this section and for Claude? Depending on whether the reader has paid for one of these services or not, navigating to chatgpt.com or claude.ai will give them access to different models.

> 2.  Rank vs. Week Plot: Each song's weekly rankings are connected with lines (`geom_line()`), and we reverse the y-axis since rank 1 is the best.
> 3.  Grouping by Track: The group aesthetic ensures lines are drawn separately for each song.
>
> **Output**
>
> This will create a line plot where each line represents a song's rank progression over weeks.
> The y-axis is reversed to reflect higher rankings at the top.

There are a few things about this answer that are promising:

-   The response is a tidyverse approach.
-   It uses `pivot_longer()` as opposed to the superseded alternative `gather()`.
-   The plot features a reversed y-axis, [just like in the book](https://r4ds.hadley.nz/data-tidy.html#fig-billboard-ranks)

However, the resulting plot doesn't show exactly what the prompt asked for -- the lines are colored by `track`, which produces a busy plot that is difficult to interpret.
While this is not ideal, it's something a new learner can easily spot and ask Chat GPT for a follow up correcting it.
For example, a subsequent prompt like the following fixes this issue:

> Can you redraw the plot without coloring each line by track?

> You can see this in action at <https://chatgpt.com/share/678f1dee-4af0-8000-befe-6df0a8d70fba>, in the second prompt.

Reviewer: My inclination here would be to just refer to this as "this saved chat" and then write "the saved chat" elsewhere and excluding the full link text, as the links included elsewhere in this section still just refer to the whole conversation.

  labs(
    title = "Song Rank vs. Week Number",
    x = "Week Number",
    y = "Rank",
    caption = "Data: Billboard (tidyr package)"
  ) +
  theme_minimal()
```

Note, however, that the only change wasn't omitting the `color = track` aesthetic mapping.
The `alpha` level is also changed (from 0.6 to 0.3) without a justification for that change and the lines are colored `"blue"`.
None of these are *bad* or *wrong* choices, but they can be confusing for new learners.
Similarly, using `theme_minimal()` is not a bad or wrong choice either[^2], but it's not *necessary*, but this might not be obvious to a new learner.

[^2]: In fact, it's my preferred ggplot2 theme!

Furthermore, while Chat GPT "solves" the problem, a thorough code review reveals a number of not-so-great things about the answer that can be confusing for new learners or promote poor coding practices:

-   The code loads packages that are not necessary: tidyr and ggplot2 packages are sufficient for this code, we don't also need dplyr. Additionally, learners coming from R for Data Science likely expect `library(tidyverse)` in analysis code, instead of loading the packages individualy.

> -   There is no need to load the `billboard` dataset, it's available to use once the tidyr package is loaded. Additionally, quotes are not needed, `data(billboard)` also works.

Reviewer: > Additionally, quotes are not needed, `data(billboard)` also works.

Just out of my curiosity—is this a "model smell" for you?

Author: Yes, I'd say so. I don't like `data()` calls when not necessary or the use of quotes when not necessary.

-   The code mixes up tidyverse and base R styles:
    -   Changing the type of `week` to numeric can be done in a `mutate()` statement with the tidyverse, which would then warrant loading the dplyr package.
    -   This can also be done within `pivot_longer()` with the `names_transform` argument.

All of these are addressable with further prompts, as I've done at <https://chatgpt.com/share/678f1dee-4af0-8000-befe-6df0a8d70fba>, in the last two prompts.
But doing so requires being able to identify these issues and explicitly asking for corrections.
In practice, I wouldn't have asked Chat GPT to correct everything for me, I would have stopped after the first suggestion, which was a pretty good starting point, and made the improvements myself.

> However, a new learner might assume (and based on my experience seeing lots of near learner code, *does* assume) the first answer is the *right* and *good* or *best* answer since (1) it looks reasonable and (2) it works, sort of.

Reviewer: ```suggestion
However, a new learner might assume (and based on my experience seeing lots of w learner code, *does* assume) the first answer is the *right* and *good* or *best* answer since (1) it looks reasonable and (2) it works, sort of.
```

2.  **Check for errors:**
    This also seems obvious -- you should run the code the tool suggests and check for errors.
    If the code gives an error, this is easy to catch and potentially easy to address.
    However, sometimes the code suggests arguments that don't exist that R might silently ignore.
    These might be unneeded arguments or a needed argument but not used properly due to how it's called or the value it's set to.
    Such errors are more difficult to identify, particularly in functions you might not be familiar with.

3.  **Run the code it gives you, line-by-line, even if the code is in a pipeline:**
    Tidyverse data transformation pipelines and ggplot layers are easy to run at once, with the code doing many things with one execution prompt, compared to Base R code where you execute each line of code separately.
    The scaffolded nature of these pipelines are very nice for keeping all steps associated with a task together and not generating unnecessary interim objects along the way.
    However, it requires self-discipline to inspect the code line-by-line as opposed to just inspecting the final output.

>     For example, I regularly encounter unnecessary `group()`/`ungroup()`s or `select()`s steps get injected into pipelines.

Reviewer: ```suggestion
    For example, I regularly encounter unnecessary `group_by()`/`ungroup()`s or `select()`s steps injected into pipelines.
```

    My recommendation would be to approach the working with AI tools for code generation with an "I'm trying to learn how to do this" attitude.
    It's then natural to investigate and interact with each step of the answer.
    If you approach it with a "Solve this for me" attitude, it's a lot harder to be critical of seemingly functioning and seemingly good enough code.

4.  **Improve code smell:**
    While I don't have empirical evidence for this, I believe for humans, taste for good code develops faster than ability.
    For LLMs, this is the opposite.
    These tools will happily barf out code that runs without regard to cohesive syntax, avoiding redundancies, etc.
    Therefore, it's essential to "clean up" the suggested code to improve its "code smell".
    Below are some steps I regularly use:
    -   Remove redundant library calls.
    -   Use `pkg::function()` syntax only as needed and consistently.
    -   Avoid mixing and matching base R and tidyverse syntax (e.g., in one step finding mean in a `summarize()` call and in another step as mean of a vector, `mean(df$var)`.
    -   Remove unnecessary `print()` statements.[^3]

> 

Reviewer: ```suggestion
    - Consider whether code comments address the "why" or the "what." If comments describe relatively self-documenting code, consider removing them.

```

However, a new learner might assume (and based on my experience seeing lots of new learner code, *does* assume) the first answer is the *right* and *good* or *best* answer since (1) it looks reasonable and (2) it works, sort of.

Furthermore, requesting improvements in subsequent calls can result in surprising changes that the user hasn't asked for.
We saw an example of this above, in updating the alpha level.
Similarly, in [the saved chat](https://chatgpt.com/share/678f1dee-4af0-8000-befe-6df0a8d70fba) you can see that asking ChatGPT to not load the packages individually but to use `library(tidyverse)` instead results in this change as well as not loading the data with a `data()` call and adding a data transformation step with `mutate()` to convert `week` to numeric:

```{r}
#| message: false
# Load the tidyverse package
library(tidyverse)

# Load the billboard dataset and prepare the data
billboard_long <- billboard |>
  pivot_longer(
    cols = starts_with("wk"),
    names_to = "week",
    values_to = "rank",
    names_prefix = "wk",
    values_drop_na = TRUE
  ) |>
  mutate(week = as.numeric(week))  # Convert week to numeric
```

> Both of these are welcome changes, but it can be surprising to a new learner why they're combined with updating the `library()` call.

Reviewer 2: I think it's worth a comment somewhere that you generally get better response when asking questions about sample datasets because the model has probably been exposed to them somewhere in their training data.

Author: Added, thanks for the suggestion

This is happening because with each call ChatGPT is "trying again" -- it's not just editing the previous answer but it's regenerating an answer with additional context.

## Case study 2: Data cleaning

The next example is a lot more vague as it's a task that, if someone asked me for help, I would reply with "Can you share the data with me?", but ChatGPT happily generates an answer.

> That's probably one of the reasons why people love these tools, they never respond to a question with another question, at least not in my experience.

Reviewer 2: I've had good luck ending my prompt with "Ask me any clarifying questions you need before continuing"

Reviewer 2: Although it didn't help here: https://claude.ai/share/57feb69a-c2f1-472a-b248-115faf02b2f5

Author: Added this in the tips section.

clubs %>% 
  head(10)
```

> This tidyverse solution:
>
> -   Uses mutate() with `if_else()` to recode the variable
> -   Uses relocate() to move membership_status to the front
> -   Shows the first 10 rows using head()
>
> The tidyverse version is more readable and follows the principle of chaining operations with the pipe operator (`%>%`).
> `if_else()` is the tidyverse equivalent of base R's `ifelse()`, with stricter type checking.
>
> Would you like me to explain any of these functions in more detail?

We're getting there!
For new learners of the tidyverse, there are still a few aspects of this code that might be misleading:

-   It doesn't use the base pipe operator (`|>`), which is the pipe operator used in R for Data Science.
-   Line breaks and indentation does not follow [tidyverse style guidelines](https://style.tidyverse.org/pipes.html#long-lines), which suggest that if the arguments to a function don't all fit on one line, each argument should go on its own line.
-   If `clubs` is a tibble, printing it will only display the first 10 rows anyway, so there's no need for the `head()` call.

Once again, it's possible to ask the tool to make these updates.

> Claude doesn't provide a sharing link for the conversation, but I've saved it at <https://gist.github.com/mine-cetinkaya-rundel/c74dc7c5f2a27a846574edd063988431>.

Reviewer 2: It does now!

Author: Ah, I probably can't recreate it, but I'll update to say it didn't at the time.

Even with vague prompts like "If clubs is a tibble, it'll only print out the first 10 rows anyway," it does exactly what we would want it to do, resulting in the following:

```{r}
#| eval: false
clubs <- clubs |>
  mutate(membership_status = if_else(is.na(membership_status), "Closed", "Open")) |>
  relocate(membership_status)

clubs
```

The operative word here being "want".
Once again, someone with enough experience with R and the tidyverse would not explicitly ask for these improvements; they would just make them themselves and carry on.
But someone who is just learning the tidyverse would likely stop much earlier.
In my experience, many new learners stop after the very first prompt, which looks unfamiliar, and hence potentially intimidating.
For learners of the tidyverse, the simple "Can you do this with the tidyverse?" can be a very useful prompt improvement that gets them majority of the way there.

## Case study 3: Web scraping

For the third and last example we'll try scraping some data from the web, specifically some [structured North Carolina weather data](https://www.usclimatedata.com/climate/durham/north-carolina/united-states/usnc0192). 

> For this example, I've used [perplexity.ai](https://www.perplexity.ai/).

Reviewer 2: Maybe mentioning why perplexity? (Presumably because it will actually download the web page?)
